keps/sig-apps/20181220-statefulset-volume-expansion.md

```diff
@@ -0,0 +1,103 @@
+  - "@kow3ns" 
+editor: TBD
+creation-date: 2018-12-20
+last-updated: 2019-06-26
+status: provisional
+see-also:
+  - https://github.com/kubernetes/enhancements/issues/531
+  - https://github.com/kubernetes/enhancements/pull/737
+  - https://github.com/kubernetes/enhancements/issues/284
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# StatefulSet Volume Expansion
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+The purpose of this enhancement is to allow for the expansion of persistent volume claims created by StatefulSets. This entails propagating increases to storage requests in `StatefulSets.volumeClaimTemplates` to associated persistent volume claims.
+
+## Motivation
+In Kubernetes v1.11 the persistent volume expansion feature was promoted to beta. This allowed users to expand volumes by editing storage requests in persistent volume claim objects.
+
+Kubernetes creates a persistent volume for each `volumeClaimTemplate` in the `volumeClaimTemplates` component of a StatefulSet. However, it is not possible to expand persistent volumes, created by StatefulSets, by editing the source `volumeClaimTemplate` in the StatefulSet object. Therefore, it is necessary for the user to individually modify all pods' persistent volume claims, by increasing their storage requests, to expand the underlying persistent volumes. This would have to be repeated each time the number of replicas in the StatefulSet object is increased, since new persistent volumes would be created with the original storage request specified in `volumeClaimTemplates` component.
+
+It would be easier and expected to allow for changes to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+Relevant Issues:
+
+* https://github.com/kubernetes/kubernetes/issues/71477
+* https://github.com/kubernetes/kubernetes/issues/72198
+
+### Goals
+
+Allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+The apiserver will allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet. Additionally, for each volumeClaimTemplate being expanded, it will be necessary to check if its associated StorageClass has volume expansion enabled. This will be achieved by using an admission controller to check if the associated StorageClasses have `allowVolumeExpansion = true`.
+
```

Comment:
Are you going to reuse existing admission controller or create a new one?

Reply:
I was just going to reuse the existing pvc/resize one.